pkg/trillianclient/trillian_client.go

```diff
@@ -220,137 +450,200 @@ func (t *TrillianClient) GetLeafAndProofByHash(ctx context.Context, hash []byte)
+				SignedLogRoot: signed,
 			},
 		}
 	}
-
 	return &Response{
-		Status: status.Code(err),
-		Err:    err,
+		Status: codes.Unknown,
+		Err:    fmt.Errorf("trillian returned empty proof for hash %s", hex.EncodeToString(hashValue)),
+	}
+}
+
+// waitForInclusionWithMinSize behaves like waitForInclusion but ensures the
+// first inclusion-proof attempt happens only after the tree has reached at
+// least minSize. This reduces initial NotFound churn without increasing time
+// to success (since inclusion requires a root advance).
+func (t *TrillianClient) waitForInclusionWithMinSize(ctx context.Context, leafHash []byte, minSize uint64) *Response {
+	start := time.Now()
+
+	// Optionally delay the very first attempt until minSize is reached.
+	// If the current snapshot is already beyond minSize, this returns immediately.
+	if err := t.waitForRootAtLeast(ctx, minSize); err != nil {
+		elapsed := float64(time.Since(start).Milliseconds())
+		metricInclusionWait.WithLabelValues("false").Observe(elapsed)
+		return &Response{Status: status.Code(err), Err: err}
+	}
+
+	for {
+		if err := ctx.Err(); err != nil {
+			elapsed := float64(time.Since(start).Milliseconds())
+			metricInclusionWait.WithLabelValues("false").Observe(elapsed)
+			return &Response{Status: status.Code(err), Err: err}
+		}
+		snap, _ := t.snapshot.Load().(rootSnapshot)
+		root := snap.root
+		signed := snap.signed
+
+		proofResp := t.getProofByHashWithRoot(ctx, leafHash, root, signed)
+		if proofResp.Err == nil || status.Code(proofResp.Err) != codes.NotFound {
+			success := proofResp.Err == nil
+			elapsed := float64(time.Since(start).Milliseconds())
+			metricInclusionWait.WithLabelValues(fmt.Sprintf("%t", success)).Observe(elapsed)
+			return proofResp
+		}
+
+		// NotFound: wait for the tree to grow and try again
+		if err := t.waitForRootAtLeast(ctx, root.TreeSize+1); err != nil {
+			return &Response{Status: status.Code(err), Err: err}
+		}
+	}
+}
+
+// waitForRootAtLeast blocks until t.lastRoot.TreeSize >= size, or context/client closes.
+func (t *TrillianClient) waitForRootAtLeast(ctx context.Context, size uint64) error {
+	start := time.Now()
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	for {
+		if err := ctx.Err(); err != nil {
+			// metrics
+			elapsed := float64(time.Since(start).Milliseconds())
+			metricWaitForRootAtLeast.WithLabelValues(fmt.Sprintf("%d", t.logID), "false").Observe(elapsed)
+			return err
+		}
+		select {
+		case <-t.stopCh:
+			elapsed := float64(time.Since(start).Milliseconds())
+			metricWaitForRootAtLeast.WithLabelValues(fmt.Sprintf("%d", t.logID), "false").Observe(elapsed)
+			return status.Error(codes.Canceled, "client closed")
+		default:
+		}
+		cur := t.snapshot.Load().(rootSnapshot)
+		if cur.root.TreeSize >= size {
+			elapsed := float64(time.Since(start).Milliseconds())
+			metricWaitForRootAtLeast.WithLabelValues(fmt.Sprintf("%d", t.logID), "true").Observe(elapsed)
+			return nil
+		}
+		t.cond.Wait()
```

Comment:
What happens when an error occurs in `updater()`? Let's say Trillian is down. From what I can tell, a request blocks until `Broadcast` is called, but `Broadcast` is only called on successful update of the tree head.

Should an error in `updater()` trigger a Broadcast so waiters unblock, and then observe an error in the context and return? Or can we have a timeout on waiting?